// ng-dynamic-component/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of ng-dynamic-component
 */
export {} from './lib/io/types';
export { defaultEventArgumentFactory, EventArgumentToken } from './lib/io/event-argument';
export { DynamicComponentInjectorToken, COMPONENT_INJECTOR } from './lib/component-injector/token';
export { ComponentOutletInjectorDirective } from './lib/component-injector/component-outlet-injector.directive';
export { ComponentOutletInjectorModule } from './lib/component-injector/component-outlet-injector.module';
export { DynamicModule } from './lib/dynamic.module';
export { DynamicComponent } from './lib/dynamic.component';
export { DynamicIoDirective } from './lib/dynamic-io/dynamic-io.directive';
export { DynamicIoModule } from './lib/dynamic-io/dynamic-io.module';
export { DynamicAttributesDirective } from './lib/dynamic-attributes/dynamic-attributes.directive';
export { DynamicAttributesModule } from './lib/dynamic-attributes/dynamic-attributes.module';
export { dynamicDirectiveDef, DynamicDirectivesDirective } from './lib/dynamic-directives/dynamic-directives.directive';
export { DynamicDirectivesModule } from './lib/dynamic-directives/dynamic-directives.module';
