// ng-dynamic-component/esm2015/lib/dynamic.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DynamicIoModule } from './dynamic-io/dynamic-io.module';
import { DynamicComponent } from './dynamic.component';
import * as ɵngcc0 from '@angular/core';
export class DynamicModule {
}
DynamicModule.ɵfac = function DynamicModule_Factory(t) { return new (t || DynamicModule)(); };
DynamicModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DynamicModule });
DynamicModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, DynamicIoModule], DynamicIoModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DynamicModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, DynamicIoModule],
                exports: [DynamicComponent, DynamicIoModule],
                declarations: [DynamicComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DynamicModule, { declarations: function () { return [DynamicComponent]; }, imports: function () { return [CommonModule, DynamicIoModule]; }, exports: function () { return [DynamicComponent, DynamicIoModule]; } }); })();
