// ng-dynamic-component/esm2015/lib/component-injector/component-outlet-injector.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-injector/component-outlet-injector.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ComponentOutletInjectorDirective } from './component-outlet-injector.directive';
import * as ɵngcc0 from '@angular/core';
export class ComponentOutletInjectorModule {
}
ComponentOutletInjectorModule.ɵfac = function ComponentOutletInjectorModule_Factory(t) { return new (t || ComponentOutletInjectorModule)(); };
ComponentOutletInjectorModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ComponentOutletInjectorModule });
ComponentOutletInjectorModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ComponentOutletInjectorModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [ComponentOutletInjectorDirective],
                declarations: [ComponentOutletInjectorDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ComponentOutletInjectorModule, { declarations: function () { return [ComponentOutletInjectorDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ComponentOutletInjectorDirective]; } }); })();
