// ng-dynamic-component/esm2015/lib/component-injector/component-outlet-injector.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-injector/component-outlet-injector.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgComponentOutlet } from '@angular/common';
import { Directive, Host } from '@angular/core';
import { DynamicComponentInjectorToken, } from './token';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
export class ComponentOutletInjectorDirective {
    /**
     * @param {?} componentOutlet
     */
    constructor(componentOutlet) {
        this.componentOutlet = componentOutlet;
    }
    /**
     * @return {?}
     */
    get componentRef() {
        // NOTE: Accessing private APIs of Angular
        return ((/** @type {?} */ (this.componentOutlet)))._componentRef;
    }
}
ComponentOutletInjectorDirective.ɵfac = function ComponentOutletInjectorDirective_Factory(t) { return new (t || ComponentOutletInjectorDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgComponentOutlet, 1)); };
ComponentOutletInjectorDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: ComponentOutletInjectorDirective, selectors: [["", "ngComponentOutlet", ""]], exportAs: ["ndcComponentOutletInjector"], features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: DynamicComponentInjectorToken,
                useExisting: ComponentOutletInjectorDirective
            },
        ])] });
/** @nocollapse */
ComponentOutletInjectorDirective.ctorParameters = () => [
    { type: NgComponentOutlet, decorators: [{ type: Host }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ComponentOutletInjectorDirective, [{
        type: Directive,
        args: [{
                // tslint:disable-next-line: directive-selector
                selector: '[ngComponentOutlet]',
                exportAs: 'ndcComponentOutletInjector',
                providers: [
                    {
                        provide: DynamicComponentInjectorToken,
                        useExisting: ComponentOutletInjectorDirective
                    },
                ]
            }]
    }], function () { return [{ type: ɵngcc1.NgComponentOutlet, decorators: [{
                type: Host
            }] }]; }, null); })();
if (false) {
    /**
     * @type {?}
     * @private
     */
    ComponentOutletInjectorDirective.prototype.componentOutlet;
}
