// ng-dynamic-component/esm2015/lib/io/io-factory.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/io/io-factory.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, Inject, Injectable, KeyValueDiffers, } from '@angular/core';
import { EventArgumentToken } from './event-argument';
import { IoService } from './io.service';
import * as i0 from "@angular/core";
import * as i1 from "./event-argument";
import * as ɵngcc0 from '@angular/core';
export class IoFactoryService {
    /**
     * @param {?} differs
     * @param {?} cfr
     * @param {?} eventArgument
     */
    constructor(differs, cfr, eventArgument) {
        this.differs = differs;
        this.cfr = cfr;
        this.eventArgument = eventArgument;
    }
    /**
     * @return {?}
     */
    create() {
        return new IoService(this.differs, this.cfr, this.eventArgument);
    }
}
IoFactoryService.ɵfac = function IoFactoryService_Factory(t) { return new (t || IoFactoryService)(ɵngcc0.ɵɵinject(ɵngcc0.KeyValueDiffers), ɵngcc0.ɵɵinject(ɵngcc0.ComponentFactoryResolver), ɵngcc0.ɵɵinject(EventArgumentToken)); };
/** @nocollapse */
IoFactoryService.ctorParameters = () => [
    { type: KeyValueDiffers },
    { type: ComponentFactoryResolver },
    { type: String, decorators: [{ type: Inject, args: [EventArgumentToken,] }] }
];
/** @nocollapse */ IoFactoryService.ɵprov = i0.ɵɵdefineInjectable({ factory: function IoFactoryService_Factory() { return new IoFactoryService(i0.ɵɵinject(i0.KeyValueDiffers), i0.ɵɵinject(i0.ComponentFactoryResolver), i0.ɵɵinject(i1.EventArgumentToken)); }, token: IoFactoryService, providedIn: "root" });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(IoFactoryService, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return [{ type: ɵngcc0.KeyValueDiffers }, { type: ɵngcc0.ComponentFactoryResolver }, { type: String, decorators: [{
                type: Inject,
                args: [EventArgumentToken]
            }] }]; }, null); })();
if (false) {
    /**
     * @type {?}
     * @private
     */
    IoFactoryService.prototype.differs;
    /**
     * @type {?}
     * @private
     */
    IoFactoryService.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    IoFactoryService.prototype.eventArgument;
}
