// ng-dynamic-component/esm2015/lib/dynamic-directives/dynamic-directives.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-directives/dynamic-directives.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ComponentOutletInjectorModule } from '../component-injector/component-outlet-injector.module';
import { DynamicDirectivesDirective } from './dynamic-directives.directive';
import * as ɵngcc0 from '@angular/core';
export class DynamicDirectivesModule {
}
DynamicDirectivesModule.ɵfac = function DynamicDirectivesModule_Factory(t) { return new (t || DynamicDirectivesModule)(); };
DynamicDirectivesModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DynamicDirectivesModule });
DynamicDirectivesModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule], ComponentOutletInjectorModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DynamicDirectivesModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [DynamicDirectivesDirective, ComponentOutletInjectorModule],
                declarations: [DynamicDirectivesDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DynamicDirectivesModule, { declarations: function () { return [DynamicDirectivesDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [DynamicDirectivesDirective, ComponentOutletInjectorModule]; } }); })();
