// ng-dynamic-component/esm2015/lib/dynamic-attributes/dynamic-attributes.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-attributes/dynamic-attributes.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ComponentOutletInjectorModule } from '../component-injector/component-outlet-injector.module';
import { DynamicAttributesDirective } from './dynamic-attributes.directive';
import * as ɵngcc0 from '@angular/core';
export class DynamicAttributesModule {
}
DynamicAttributesModule.ɵfac = function DynamicAttributesModule_Factory(t) { return new (t || DynamicAttributesModule)(); };
DynamicAttributesModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DynamicAttributesModule });
DynamicAttributesModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule], ComponentOutletInjectorModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DynamicAttributesModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [DynamicAttributesDirective, ComponentOutletInjectorModule],
                declarations: [DynamicAttributesDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DynamicAttributesModule, { declarations: function () { return [DynamicAttributesDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [DynamicAttributesDirective, ComponentOutletInjectorModule]; } }); })();
