// ng2-material-dropdown/esm5/src/modules/services/dropdown-state.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Ng2DropdownState } from './ng2-dropdown-state';
import * as ɵngcc0 from '@angular/core';
var DropdownStateService = /** @class */ (function () {
    function DropdownStateService() {
        this.menuState = {
            isVisible: false,
            toString: function () {
                return this.isVisible === true ? 'visible' : 'hidden';
            }
        };
        this.dropdownState = new Ng2DropdownState();
    }
DropdownStateService.ɵfac = function DropdownStateService_Factory(t) { return new (t || DropdownStateService)(); };
DropdownStateService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DropdownStateService, factory: function (t) { return DropdownStateService.ɵfac(t); } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DropdownStateService, [{
        type: Injectable
    }], function () { return []; }, null); })();
    return DropdownStateService;
}());
export { DropdownStateService };
