// bn-ng-idle/esm5/lib/bn-ng-idle.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { BnNgIdleService } from './bn-ng-idle.service';
import * as ɵngcc0 from '@angular/core';
var BnNgIdleModule = /** @class */ (function () {
    function BnNgIdleModule() {
    }
BnNgIdleModule.ɵfac = function BnNgIdleModule_Factory(t) { return new (t || BnNgIdleModule)(); };
BnNgIdleModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BnNgIdleModule });
BnNgIdleModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [BnNgIdleService], imports: [[]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BnNgIdleModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [],
                providers: [BnNgIdleService],
                exports: []
            }]
    }], function () { return []; }, null); })();
    return BnNgIdleModule;
}());
export { BnNgIdleModule };
