// ng-dynamic-component/esm2015/lib/dynamic-io/dynamic-io.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic-io/dynamic-io.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ComponentOutletInjectorModule } from '../component-injector/component-outlet-injector.module';
import { DynamicIoDirective } from './dynamic-io.directive';
import * as ɵngcc0 from '@angular/core';
export class DynamicIoModule {
}
DynamicIoModule.ɵfac = function DynamicIoModule_Factory(t) { return new (t || DynamicIoModule)(); };
DynamicIoModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DynamicIoModule });
DynamicIoModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule], ComponentOutletInjectorModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DynamicIoModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [DynamicIoDirective, ComponentOutletInjectorModule],
                declarations: [DynamicIoDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DynamicIoModule, { declarations: function () { return [DynamicIoDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [DynamicIoDirective, ComponentOutletInjectorModule]; } }); })();
